// pptxgenjs/index.js
import fs from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import { spawnSync } from 'node:child_process';

function xmlEscape(value = '') {
  return String(value)
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&apos;');
}

class Slide {
  constructor() {
    this.items = [];
    this.background = { color: 'FFFFFF' };
  }

  addText(text, options = {}) {
    this.items.push({ type: 'text', text, options });
  }

  addShape(shape, options = {}) {
    this.items.push({ type: 'shape', shape, options });
  }
}

export default class PptxGenJS {
  static ShapeType = {
    rect: 'rect',
    roundRect: 'roundRect',
  };

  constructor() {
    this.slides = [];
    this.layout = 'LAYOUT_WIDE';
    this.author = '';
    this.company = '';
    this.subject = '';
    this.title = '';
    this.lang = 'en-US';
    this.theme = {};
  }

  get ShapeType() {
    return PptxGenJS.ShapeType;
  }

  addSlide() {
    const slide = new Slide();
    this.slides.push(slide);
    return slide;
  }

  async writeFile({ fileName }) {
    const tempRoot = await fs.mkdtemp(path.join(os.tmpdir(), 'pptxgenjs-local-'));
    const dirs = [
      '_rels',
      'docProps',
      'ppt',
      'ppt/_rels',
      'ppt/slides',
      'ppt/slides/_rels',
      'ppt/theme',
    ];

    await Promise.all(dirs.map((dir) => fs.mkdir(path.join(tempRoot, dir), { recursive: true })));

    await fs.writeFile(path.join(tempRoot, '[Content_Types].xml'), contentTypesXml(this.slides.length));
    await fs.writeFile(path.join(tempRoot, '_rels', '.rels'), rootRelsXml());
    await fs.writeFile(path.join(tempRoot, 'docProps', 'core.xml'), coreXml(this));
    await fs.writeFile(path.join(tempRoot, 'docProps', 'app.xml'), appXml(this.slides.length));
    await fs.writeFile(path.join(tempRoot, 'ppt', 'presentation.xml'), presentationXml(this.slides.length));
    await fs.writeFile(path.join(tempRoot, 'ppt', '_rels', 'presentation.xml.rels'), presentationRelsXml(this.slides.length));
    await fs.writeFile(path.join(tempRoot, 'ppt', 'theme', 'theme1.xml'), themeXml());

    await Promise.all(this.slides.map((slide, index) => (
      fs.writeFile(path.join(tempRoot, 'ppt', 'slides', `slide${index + 1}.xml`), slideXml(slide, index + 1))
    )));

    const python = spawnSync('python3', ['-c', ZIP_SCRIPT, tempRoot, fileName], { encoding: 'utf8' });
    if (python.status !== 0) {
      throw new Error(`Failed to package pptx: ${python.stderr || python.stdout}`);
    }
  }
}

const ZIP_SCRIPT = String.raw`
import os, sys, zipfile
src, dest = sys.argv[1], sys.argv[2]
with zipfile.ZipFile(dest, 'w', compression=zipfile.ZIP_DEFLATED) as zf:
    for root, _, files in os.walk(src):
        for name in files:
            full = os.path.join(root, name)
            rel = os.path.relpath(full, src)
            zf.write(full, rel)
`;

function contentTypesXml(slideCount) {
  const slideOverrides = Array.from({ length: slideCount }, (_, index) =>
    `<Override PartName="/ppt/slides/slide${index + 1}.xml" ContentType="application/vnd.openxmlformats-officedocument.presentationml.slide+xml"/>`).join('');
  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
  <Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>
  <Default Extension="xml" ContentType="application/xml"/>
  <Override PartName="/docProps/app.xml" ContentType="application/vnd.openxmlformats-officedocument.extended-properties+xml"/>
  <Override PartName="/docProps/core.xml" ContentType="application/vnd.openxmlformats-package.core-properties+xml"/>
  <Override PartName="/ppt/presentation.xml" ContentType="application/vnd.openxmlformats-officedocument.presentationml.presentation.main+xml"/>
  <Override PartName="/ppt/theme/theme1.xml" ContentType="application/vnd.openxmlformats-officedocument.theme+xml"/>
  ${slideOverrides}
</Types>`;
}

function rootRelsXml() {
  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
  <Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="ppt/presentation.xml"/>
  <Relationship Id="rId2" Type="http://schemas.openxmlformats.org/package/2006/relationships/metadata/core-properties" Target="docProps/core.xml"/>
  <Relationship Id="rId3" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/extended-properties" Target="docProps/app.xml"/>
</Relationships>`;
}

function coreXml(doc) {
  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<cp:coreProperties xmlns:cp="http://schemas.openxmlformats.org/package/2006/metadata/core-properties" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:dcmitype="http://purl.org/dc/dcmitype/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <dc:title>${xmlEscape(doc.title)}</dc:title>
  <dc:subject>${xmlEscape(doc.subject)}</dc:subject>
  <dc:creator>${xmlEscape(doc.author)}</dc:creator>
  <cp:lastModifiedBy>${xmlEscape(doc.author)}</cp:lastModifiedBy>
</cp:coreProperties>`;
}

function appXml(slideCount) {
  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Properties xmlns="http://schemas.openxmlformats.org/officeDocument/2006/extended-properties" xmlns:vt="http://schemas.openxmlformats.org/officeDocument/2006/docPropsVTypes">
  <Application>pptx-agent</Application>
  <Slides>${slideCount}</Slides>
</Properties>`;
}

function presentationXml(slideCount) {
  const slideRefs = Array.from({ length: slideCount }, (_, index) =>
    `<p:sldId id="${256 + index}" r:id="rId${index + 1}"/>`).join('');
  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<p:presentation xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:p="http://schemas.openxmlformats.org/presentationml/2006/main">
  <p:sldMasterIdLst/>
  <p:sldIdLst>${slideRefs}</p:sldIdLst>
  <p:sldSz cx="12192000" cy="6858000"/>
  <p:notesSz cx="6858000" cy="9144000"/>
</p:presentation>`;
}

function presentationRelsXml(slideCount) {
  const rels = Array.from({ length: slideCount }, (_, index) =>
    `<Relationship Id="rId${index + 1}" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/slide" Target="slides/slide${index + 1}.xml"/>`).join('');
  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">${rels}</Relationships>`;
}

function themeXml() {
  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<a:theme xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main" name="Office Theme">
  <a:themeElements>
    <a:clrScheme name="Custom"><a:dk1><a:srgbClr val="000000"/></a:dk1><a:lt1><a:srgbClr val="FFFFFF"/></a:lt1><a:accent1><a:srgbClr val="2563EB"/></a:accent1></a:clrScheme>
    <a:fontScheme name="Custom"><a:majorFont/><a:minorFont/></a:fontScheme>
    <a:fmtScheme name="Custom"><a:fillStyleLst/><a:lnStyleLst/><a:effectStyleLst/><a:bgFillStyleLst/></a:fmtScheme>
  </a:themeElements>
</a:theme>`;
}

function normalizeText(text) {
  if (Array.isArray(text)) {
    return text.map((part) => typeof part === 'string' ? part : part.text).join('\n');
  }
  return String(text ?? '');
}

function slideXml(slide, index) {
  const content = slide.items
    .filter((item) => item.type === 'text')
    .map((item, itemIndex) => {
      const text = normalizeText(item.text)
        .split(/\n+/)
        .map((paragraph) => `<a:p><a:r><a:t>${xmlEscape(paragraph)}</a:t></a:r></a:p>`)
        .join('');
      return `<p:sp>
        <p:nvSpPr><p:cNvPr id="${itemIndex + 2}" name="TextBox ${itemIndex + 1}"/><p:cNvSpPr txBox="1"/><p:nvPr/></p:nvSpPr>
        <p:spPr/>
        <p:txBody><a:bodyPr/><a:lstStyle/>${text}</p:txBody>
      </p:sp>`;
    })
    .join('');

  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<p:sld xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:p="http://schemas.openxmlformats.org/presentationml/2006/main">
  <p:cSld name="Slide ${index}">
    <p:spTree>
      <p:nvGrpSpPr><p:cNvPr id="1" name=""/><p:cNvGrpSpPr/><p:nvPr/></p:nvGrpSpPr>
      <p:grpSpPr/>
      ${content}
    </p:spTree>
  </p:cSld>
  <p:clrMapOvr><a:masterClrMapping/></p:clrMapOvr>
</p:sld>`;
}
